### PR TITLE
[SPARC] Prefer RDPC over CALL to implement GETPCX for 64-bit target

### PR DESCRIPTION
**Pull request by koachan**
On 64-bit target, prefer usng RDPC over CALL to get the value of %pc.
This is faster on modern processors (Niagara T1 and newer) and avoids polluting
the processor's predictor state.

The old behavior of using a fake CALL is still done when tuning for classic
UltraSPARC processors, since RDPC is much slower there.

A quick pgbench test on a SPARC T4 shows about 2% speedup on SELECT loads,
and about 7% speedup on INSERT/UPDATE loads.

### CODE REVIEW

File: llvm/test/CodeGen/SPARC/tune-getpcx.ll [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+@value = external global i32
+
+; CALL: call
+; CALL-NOT: rd %pc
+; RDPC: rd %pc
+; RDPC-not: call
+define i32 @test() {
+  %1 = load i32, i32* @value
+  ret i32 %1
+}
```
Comment (s-barannikov) [inline llvm/test/CodeGen/SPARC/tune-getpcx.ll:18]: The test should show the full expansion (autogenerate the checks?).

File: llvm/lib/Target/Sparc/SparcAsmPrinter.cpp [hunk 2]
```diff
@@ -234,8 +244,15 @@ void SparcAsmPrinter::LowerGETPCXAndEmitMCInsts(const MachineInstr *MI,
...
   OutStreamer->emitLabel(StartLabel);
-  MCOperand Callee =  createPCXCallOP(EndLabel, OutContext);
-  EmitCall(*OutStreamer, Callee, STI);
+  if (!STI.getTargetTriple().isSPARC64() ||
+      STI.hasFeature(Sparc::TuneSlowRDPC)) {
+    MCOperand Callee = createPCXCallOP(EndLabel, OutContext);
+    EmitCall(*OutStreamer, Callee, STI);
+  } else {
+    // TODO make it possible to store PC in other registers
+    // so that leaf function optimization becomes possible.
```
Comment (s-barannikov) [inline llvm/lib/Target/Sparc/SparcAsmPrinter.cpp:253]: Would that be hard to do the expansion at instruction selection level?

File: llvm/lib/Target/Sparc/SparcAsmPrinter.cpp [hunk 3]
```diff
@@ -234,8 +244,15 @@ void SparcAsmPrinter::LowerGETPCXAndEmitMCInsts(const MachineInstr *MI,
   //   add <MO>, %o7, <MO>
```
Comment (s-barannikov) [inline llvm/lib/Target/Sparc/SparcAsmPrinter.cpp:244]: Please update the comment to include the rdpc case.

File: llvm/lib/Target/Sparc/Sparc.td [hunk 4]
```diff
@@ -118,9 +126,11 @@ def : Proc<"ma2480",          [FeatureLeon, LeonCASA]>;
 def : Proc<"ma2485",          [FeatureLeon, LeonCASA]>;
 def : Proc<"ma2x8x",          [FeatureLeon, LeonCASA]>;
 def : Proc<"v9",              [FeatureV9]>;
-def : Proc<"ultrasparc",      [FeatureV9, FeatureV8Deprecated, FeatureVIS]>;
+def : Proc<"ultrasparc",      [FeatureV9, FeatureV8Deprecated, FeatureVIS],
+                              [TuneSlowRDPC]>;
```
Comment (s-barannikov) [inline llvm/lib/Target/Sparc/Sparc.td:130]: Why is it in TuneFeatures? Sparc doesn't seem to support -mtune.

Comment (s-barannikov) [reply llvm/lib/Target/Sparc/Sparc.td:130]: I see. IIUC -mtune is translate into -tune-cpu llc option. If so, the test should use this option.

File: llvm/test/CodeGen/SPARC/getpcx-call.ll [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+;; SPARC32 and SPARC64 for classic UltraSPARCs implement GETPCX
+;; with a fake `call`.
+;; All other SPARC64 targets implement it with `rd %pc, %o7`.
+;; Need to do the tests in separate files because apparently `tune-cpu`
+;; attribute applies to the entire file at once.
+
+@value = external global i32
+
+define i32 @testCall() #0 {
```
Comment (s-barannikov) [inline llvm/test/CodeGen/SPARC/getpcx-call.ll:13]:
```suggestion
define i32 @testCall() nounwind #0 {
```
to eliminate irrelevant cfi directives

File: llvm/lib/Target/Sparc/SparcAsmPrinter.cpp (hunk 2 above)
Comment (s-barannikov) [reply llvm/lib/Target/Sparc/SparcAsmPrinter.cpp]: Let's do it separately then.

File: llvm/test/CodeGen/SPARC/getpcx-call.ll [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+; SPARC64-NEXT:  .Ltmp1:
+; SPARC64-NEXT:    or %i0, %lo(_GLOBAL_OFFSET_TABLE_+(.Ltmp1-.Ltmp0)), %i0
+; SPARC64-NEXT:    add %i0, %o7, %i0
+; SPARC64-NEXT:    sethi %hi(value), %i1
+; SPARC64-NEXT:    add %i1, %lo(value), %i1
+; SPARC64-NEXT:    ldx [%i0+%i1], %i0
+; SPARC64-NEXT:    ld [%i0], %i0
+; SPARC64-NEXT:    ret
+; SPARC64-NEXT:    restore
+  %1 = load i32, i32* @value
```
Comment (s-barannikov) [inline llvm/test/CodeGen/SPARC/getpcx-call.ll:47]:
```suggestion
  %1 = load i32, ptr @value
```
I thought it does not compile anymore :thinking:

**Review by s-barannikov — APPROVED** [review]

**Comment by brad0** [comment]
I intentionally restored the branch as I only noticed after the merge that the target branch was set incorrectly. It wasn't merged into the LLVM repo.

**Comment by s-barannikov** [comment]
Yes, sure

**Comment by brad0** [comment]
> Uh oh, seems like I couldn't merge it with the tool either... Guess I'll open a new PR to merge this, would that be okay? @brad0 @s-barannikov

Ya, sure. Whatever path is easier and quickest for you.

**Pull request merged**